### DISCUSSION TITLE
SQL support

### DISCUSSION THREAD

**Discussion by shawncao** [varchar-io/nebula#60] [General]
Hi @chenqin, your recent lib_pg parser makes me recall some thoughts on SQL support before, citing here for discussion - 
There is a project created long time ago talking about SQL support - https://github.com/varchar-io/nebula/projects/11

**Comment by chenqin** [comment]
@shawncao The context were I tried to figure out which is most **intuitive way requiring less to learn** to move data into nebula system without learning many own syntax and definitions and build layers of custom code. 
To me, I think if I were give standard sql statement (PostgreSQL parser) and build a "materialized view" in nebula in realtime sounds easier than checking ingestion spec (which I doubt if majority of user bother to check before things stop working)

I also know from user query prospective, we might just do UI and "dataframe" API like to accommodate complex query requirement. 

@shawncao what do you think.

**Reply by shawncao** [reply]
first of all, I'm not a fan of SQL, the biggest reason is it is hard to express hard logic (especially transformations) in SQL without flexible /easy approach to support UDF.

But I do think from user's perspective, SQL is the best way to standardize interface to allow user to pick up soon. I cited the SQL project to you is to see if you can think about them as holistically rather than just design it for ingestion only.  IMO, introducing SQL language for ingestion part only may be either overkill or hard to address all different kinds of needs. But I support you if you want to experiment with it anyways.

Specifically for ingestion, IMO, if we're not talking about SQL, probably JS is a good idea to wrap user's transformer logic and sampling logic.

**Reply by chenqin** [reply]
I see, if we go with js route, it means we need to port reader and parser to node js and pass to c++ layer. Do you have sample code I can evaluate at same time? 

One gap I observed is Druid to some extent Rocksdb cluster are serving as "Serving" layer at the end of data system lifecyle. And it seems lack of friendly and easy way to get formatted structure data into system easily. 

- If topic is too large, they can't do stream to table window join before landing to druid table
- if data needs to enrich, they can't carry RPC to services and get results (async UDF)
- if data needs to combine multiple columns, it seems took quite a bit learning their own spec standard

Either node.js or SQL can both do resolve as long as we have UDFs in place. We can evaluate both options in next couple of months and see which works best for prospective users.

**Comment by shawncao** [comment]
It would be helpful if we start with a concrete scenario, and mock the
ingestion spec with both SQL and JS and see how they look like and have
some discussions from there.

**Reply by chenqin** [reply]
sounds good, will do

**Comment by chenqin** [comment]
track ongoing evaluation https://github.com/varchar-io/nebula/pull/59
